# Task 28-> Exploring Cross-Validation, Overfitting, and Underfitting

#######
Cross-Validation
Cross-validation is a technique used to assess the performance and generalization ability of a model by partitioning the dataset into multiple subsets or folds.
Overfitting
Overfitting occurs when a model learns the details and noise in the training data to the extent that it negatively impacts its performance on new, unseen data. Characteristics include:
Underfitting
Underfitting occurs when a model is too simple to capture the underlying patterns in the data. 
Characteristics include:

### Import necessary libraries and dataset

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import LabelEncoder

file_path = r'C:\Users\Huawei\Desktop\Titanic-Dataset.csv'
data = pd.read_csv(file_path)

### Split data into features and target

In [2]:
data['Sex'] = LabelEncoder().fit_transform(data['Sex'])  # Convert 'Sex' to numeric to remove error cannot convert string to float 'male'

# Split data into features and target
X = data[['Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare']]
y = data['Survived']


### Split data into training and testing sets

In [3]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

### Initialize models

In [4]:
logistic_model = LogisticRegression(max_iter=1000)
tree_model = DecisionTreeClassifier(max_depth=2) # Simple model (underfitting)
complex_tree_model = DecisionTreeClassifier(max_depth=10) # Complex model (overfitting)

### Cross-Validation

In [5]:
def evaluate_model(model, X, y):
    scores = cross_val_score(model, X, y, cv=5, scoring='accuracy')
    return np.mean(scores)

### Evaluate models

In [6]:
print("Logistic Model CV Score:", evaluate_model(logistic_model, X_train, y_train)*100)
print("Simple Tree Model CV Score:", evaluate_model(tree_model, X_train, y_train)*100)
print("Complex Tree Model CV Score:", evaluate_model(complex_tree_model, X_train, y_train)*100)

Logistic Model CV Score: 66.0
Simple Tree Model CV Score: 78.0
Complex Tree Model CV Score: 86.99999999999999


### My Understanding Of the CV Score
### . As CV(Cross Validation) score tells us that how well the model will perform on new/unseen data. 
### . So logistic regression model correctly predicts the survival status about 66% of the time, which is relatively low as compare to simple tree model which have cv score of 78%.
### . Complex tree model has the highest CV Score among all models used with approx 87% of the correct prediction about survival status.

### My understanding Of Overfitting
### . Overfitting occurs when a model learns irrrelevant data during training to the extent that it negatively impacts its performance on new, unseen data. 
### . When there is less training data, model learns irreleveant data rather than general pattrens which results to lower accuracy on unseen data despite having high accuracy on training data.
### . In above code in complex_tree_model, max_depth(no. of levels tree can have)=10, if I increase it further, it will capture more complex pattrens but can lead to overfitting.


### My understanding of Underfitting
### . Underfitting happens when a model is too basic to understand the patterns in the data.
### . As a result, it performs poorly on both the training data and new data. Thus It leads to lower accuracy.
### . In above code in tree model, max_depth=2, so decreasing max_depth further simplifies the model, thus reducing the risk of overfitting but it can lead to underfitting if too low.